### PR TITLE
feat: add support for field metadata

### PR DESCRIPTION
**Pull request by khaledh**
@mattjw 👋🏼 It's me again :)

This PR adds support for field metadata, which is [natively supported](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.types.StructField.html) by Spark, so it just passes it through for the most part.

Use case: we're using this feature to support BigQuery column-level policy tags, which we stash in Spark field metadata. Another use case is column descriptions.

### CODE REVIEW

**Comment by mattjw** [comment]
Sorry for the delay @khaledh. I will take a loo shortly 👀 .

**Review by mattjw — COMMENTED** [review]
Thank you for the PR @khaledh! Great to see more and more of pyspark's functionality getting supported in sparkql.

Two final things, but other than that LGTM ❤️ :

- I added a question about merging metadata and precedence. (I don't have a strong opinion – happy to take your insight as a regular pyspark user!)
- There's some test coverage to add (build step [here](https://app.circleci.com/pipelines/github/mattjw/sparkql/383/workflows/5461a9ee-bde1-4d9f-b8ce-2f88e21dad9f/jobs/413/parallel-runs/0/steps/0-105) an report [here](https://output.circle-artifacts.com/output/job/fcc55fef-40c2-411e-b2eb-9a056ae70052/artifacts/0/reports/htmlcov/index.html)). I'll have a bit of time tomorrow to mop this up myself 🧹. I'll add it to the PR.

File: sparkql/schema_merger.py [hunk 1]
```diff
@@ -84,6 +84,7 @@ def merge_fields(cls, field_a: StructField, field_b: StructField) -> StructField
             name=field_a.name,
             dataType=cls.merge_types(field_a.dataType, field_b.dataType, parent_field_name=field_a.name),
             nullable=field_a.nullable,
+            metadata={**(field_a.metadata or {}), **(field_b.metadata or {})},
```
[the remark this answers is not in the document]
Comment (mattjw) [reply sparkql/schema_merger.py:87]: 📝 not to self: I'll add a test to cover the non-commutativity.

[the remark this answers is not in the document]
Comment (khaledh) [reply sparkql/schema_merger.py:87]: Good point. I think we should fail if there are matching keys. The expectation is that merging shouldn't silently let the "last one wins". I can make that change.

**Comment by mattjw** [comment]
📝 Note for posterity. I went down a rabbit hole investigating how PySpark factors `metadata` into comparison and equality. Answers, by way of example...

```
>>> {StructType([StructField("some_field", StringType(), metadata=None)]) == StructType([StructField("some_field", StringType(), metadata={})])}
True

>>> {StructType([StructField("some_field", StringType(), metadata={"a": "b"})]) == StructType([StructField("some_field", StringType(), metadata={})])}
False
```

So in summary, different metadata can render two schemas inequal. Which is sensible. And PySpark treats None as {}. (I think behind the scenes it's just replacing the None it receives in the constructor with a {}.) I'll replicate the same behaviour in PySpark.

File: sparkql/schema_merger.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattjw) [reply sparkql/schema_merger.py:99]: Thank you for the revision. The new logic makes sense to me.

**Review by mattjw — COMMENTED** [review]
Thanks for updating the merging logic, @khaledh . And thanks again for the PR. Sounds like sparkql is getting some use! Great to hear it's useful to some folks out there.

This is LGTM now. I made a few small changes:

- Additional test coverage. 
- Treat `metadata=None` as `{}`.This seems sensible to me, since [that's what pyspark seems to do](https://github.com/mattjw/sparkql/pull/113#issuecomment-1656328048). If see any unwanted consequences of that, please let me know.
- Edited the PR title to `feat: xyz`, to conform with this project's [conventional commits convention](https://github.com/mattjw/sparkql/blob/master/CONTRIBUTING.md#versioning-and-tracking-changes). (The build tooling also uses this to decide version bumps.)

Please feel free to comment or critique any of those changes or recent comments, if you think they could be problematic – feedback welcome!

I'll merge later this weekend and publish a release.

**Comment by khaledh** [comment]
Amazing! Thanks @mattjw for your continued support on sparkql 🙏🏼 It's being put to good use :)

**Pull request merged**